SeaBattle/Ships/BaseShip.cs

```diff
@@ -0,0 +1,159 @@
+        }
+        public override bool Equals(object s1) => Equals(s1 as BaseShip);
+        // To do in the MoveTo method  
+        // 1. + calculating mid-ship coordinates of current position
+        // 2. + calculating mid-ship coordinates of new position
+        // 3. + calcualating distance of the move
+        // 4. + checking if speed is enough to get to new position
+        // 5. + checking if new position is available
+        // 6. + updating current position with new position
+        public void MoveTo(List<Point> NewPosition, Sea Sea1)
+        {
+            float midShipX = 0;
+            float midShipY = 0;
+            foreach (var d in Decks)
+            {
+                midShipX += d.Location.X;
+                midShipY += d.Location.Y;
+            }
+            midShipX /= Decks.Count;
+            midShipY /= Decks.Count;
+
+            float midNewX = 0;
+            float midNewY = 0;
+            foreach (var p in NewPosition)
+            {
+                midNewX += p.X;
+                midNewY += p.Y;
+            }
+            midNewX /= NewPosition.Count;
+            midNewY /= NewPosition.Count;
+
+            float TravelX = (float)Math.Pow(midShipX - midNewX, 2);
+            float TravelY = (float)Math.Pow(midShipY - midNewY, 2);
+            float TravelDistance;
+            TravelDistance = (float)Math.Sqrt(TravelX + TravelY);
+            if (Speed < TravelDistance)
+            {
+                throw new Exception("Ship can't move");
+            }
+            if (Sea1.IsPlaceAvailable(this) && this.ShipCoordsLine(NewPosition) && this.ShipCoordsUnion(NewPosition) && this.ShipCoordsUnique(NewPosition))
+            {
+                for (int i = 0; i < NewPosition.Count; i++)
+                {
+                    Decks[i].Location.X = NewPosition[i].X;
+                    Decks[i].Location.Y = NewPosition[i].Y;
+                }
+            }
+            else
+                throw new Exception("Ship can't move");
+        }
+        public bool HasCoordinatesIntersect(List<Deck> DecksList)
+        {
+            return Decks.Any(e => e.HasCoordinatesIntersect(DecksList));
+        }
+        public bool ShipCoordsLine(List<Point> SameDecks)
+        {
+            bool lineFlagX = true;
+            bool lineFlagY = true;
+
+            for (int i = 0; i < SameDecks.Count - 1; i++)
+            {
+                lineFlagX &= SameDecks[i].X == SameDecks[i + 1].X;
+                lineFlagY &= SameDecks[i].Y == SameDecks[i + 1].Y;
+            }
+            return (lineFlagX ^ lineFlagY);
+        }
+        public bool ShipCoordsUnion(List<Point> SameDecks)
+        {
+            bool unionFlagX = true;
+            bool unionFlagY = true;
+            for (int i = 0; i < SameDecks.Count - 1; i++)
+            {
+                unionFlagX &= (SameDecks[i].X == SameDecks[i + 1].X) || (SameDecks[i].X == SameDecks[i + 1].X + 1) || (SameDecks[i].X == SameDecks[i + 1].X - 1);
```

Comment:
By the way what is the difference between & and && operators?

Reply:
The difference is & is a bitwise operator and && is logical, && does not work on non-logical types and it is short-circuit logical operator

Reply:
It's not a bitwise operator in that case, it can be a logical operator as well if applied to the boolean values, but what I wanted to hear is the difference between the short circuit operator and the default one.

Reply:
Guess in Skype was exact answer you wanted to see.

Reply:
Yep